### PR TITLE
Get token to search V2 registries without creds

### PR DESCRIPTION
**Pull request by umohnani8**
Update the search functionality to be able to search registries
with a V2 endpoint by generating a token.

Signed-off-by: umohnani8 <umohnani@redhat.com>

### CODE REVIEW

File: docker/docker_client.go [hunk 1]
```diff
@@ -292,7 +293,7 @@ func SearchRegistry(ctx context.Context, sys *types.SystemContext, registry, ima
 		registry = dockerV1Hostname
 	}
 
-	client, err := newDockerClientWithDetails(sys, registry, username, password, "", nil, "")
+	client, err := newDockerClientWithDetails(sys, registry, username, password, "*", nil, registry)
```
Comment (umohnani8) [inline docker/docker_client.go]: @mtrmac what is the best way of generating the `remoteName` from the registry given by the user?

File: docker/docker_client.go [hunk 2]
```diff
@@ -479,12 +480,13 @@ func (c *dockerClient) getBearerToken(ctx context.Context, realm, service, scope
...
 	}
-	tokenBlob, err := ioutil.ReadAll(res.Body)
+	tokenBlob := new(bytes.Buffer)
+	_, err := tokenBlob.ReadFrom(res.Body)
 	if err != nil {
 		return nil, err
 	}
 
-	return newBearerTokenFromJSONBlob(tokenBlob)
+	return newBearerTokenFromJSONBlob(tokenBlob.Bytes())
```
Comment (mtrmac) [inline docker/docker_client.go]: What does this change do? `ioutil.ReadAll` internally uses a `bytes.Buffer` in a very similar way.

File: docker/docker_client.go (hunk 1 above)
Comment (mtrmac) [reply docker/docker_client.go]:
```console
$ curl -i  https://registry-1.docker.io/v2/_catalog
HTTP/1.1 401 Unauthorized
…
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="registry:catalog:*"
…
```

Also looking at https://github.com/docker/distribution/blob/4a75b72fd3621c1892143614474ff2b05d30e4b3/registry/handlers/app.go#L942-L953 , `registry:catalog:*` seems to be the scope we want.

The current `remoteName`/`actions` format can not produce a `scope="registry:…`, so the `newDockerClientWithDetails` API will need to be changed.

Maybe, to reduce the number of unnamed parameters, make `authScope` a struct of `type, name, actions string` (to follow https://github.com/docker/distribution/blob/master/docs/spec/auth/scope.md), and have a pass a single `authScope` parameter to `newDockerClientWithDetails`.  (Or there could be a single `scope string` parameter, but the callers would have to worry about formatting; well-typed data concentrates that into a single place.)

**Comment by mtrmac** [comment]
> make `authScope` a struct of `type, name, actions string`

Or maybe even `type, name string; actions []string`, to be extra correct.

File: docker/docker_client.go (hunk 1 above)
Comment (mtrmac) [reply docker/docker_client.go]: > make `authScope` a struct of `type, name, actions string`

Or maybe even `type, name string; actions []string`, to be extra correct.

File: docker/docker_client.go (hunk 2 above)
Comment (umohnani8) [reply docker/docker_client.go]: @mtrmac it was breaking and throwing an error "unknown end of JSON input" and when I changed it to read in the token like this, it stopped throwing the error.

Comment (mtrmac) [reply docker/docker_client.go]: That’s surprising.  Can you compare the two values passed to `newBearerTokenFromJSONBlob`?

**Comment by umohnani8** [comment]
Closed in favour of https://github.com/containers/image/pull/502

**Pull request closed without merging**